File: src/server/nodes/fix-query.ts
/**
 * LLM-based SQL fix node — attempts to repair a failed SQL query.
 */

import { generateText } from 'ai';
import { openai } from '@ai-sdk/openai';
import type { PipelineState } from './types';
import { LLM_MODEL_NAME } from '../config';

export interface FixQueryResult {
  generatedQuery: string;
  querySource: 'fallback';
  errorMessage: string | null;
}

export async function fixQuery(state: PipelineState): Promise<FixQueryResult> {
  const { generatedQuery, errorMessage, selectedTable, tableSchema, merchantId, userPrompt } =
    state;

  console.log(`[fixQuery] Attempting to fix query for table: ${selectedTable}`);
  console.log(`[fixQuery] Error was: ${errorMessage}`);

  const systemPrompt = `You are a MySQL query repair expert. A query failed with an error. Fix it.

Table: ${selectedTable}
Schema: ${tableSchema}
Merchant ID: ${merchantId}

Original query:
${generatedQuery}

Error:
${errorMessage}

Rules:
- Always include WHERE idMerchant = ${merchantId}
- Only use SELECT (no INSERT/UPDATE/DELETE)
- Use backticks for column and table names
- Add LIMIT 100 if no limit specified
- Return only the corrected SQL query, no explanation`;

  try {
    const { text } = await generateText({
      model: openai(LLM_MODEL_NAME),
      messages: [
        { role: 'system', content: systemPrompt },
        {
          role: 'user',
          content: `Fix this query for the user's question: "${userPrompt}"`,
        },
      ],
    });

    const fixedQuery = extractSql(text);
    console.log(`[fixQuery] Fixed query: ${fixedQuery.slice(0, 200)}`);

    return {
      generatedQuery: fixedQuery,
      querySource: 'fallback',
      errorMessage: null,
    };
  } catch (err) {
    const errMsg = err instanceof Error ? err.message : String(err);
    console.error(`[fixQuery] LLM fix failed: ${errMsg}`);
    return {
      generatedQuery: generatedQuery,
      querySource: 'fallback',
      errorMessage: `Fix attempt failed: ${errMsg}`,
    };
  }
}

function extractSql(text: string): string {
  const codeBlockMatch = text.match(/```(?:sql)?\s*\n?([\s\S]*?)\n?```/);
  if (codeBlockMatch) return codeBlockMatch[1].trim();

  const sqlMatch = text.match(/SELECT[\s\S]*/i);
  if (sqlMatch) return sqlMatch[0].trim();

  return text.trim();
}

File: agent/nodes/fix_query.py
import logging
from langchain_core.prompts import ChatPromptTemplate
from langchain_core.output_parsers.string import StrOutputParser
from agent.config import get_llm
from .types import State

def fix_query(state: State) -> dict:
    """Analyzes a failed SQL query and attempts to fix it."""
    logging.info("--- Fixing SQL Query ---")
    
    original_query = state["generated_query"]
    error_message = state["error_message"]
    table_schema = state["table_schema"]
    
    prompt = ChatPromptTemplate.from_template(
        """Fix this failed MySQL query. Return ONLY the corrected SQL.

**Schema:**
```sql
{table_schema}
```

**Failed Query:**
```sql
{original_query}
```

**Error:** {error_message}

**Rules:** Use only the table above. Enclose identifiers in backticks. If table is missing, return: `SELECT 'Table not available' AS error`

**Corrected SQL:**"""
    )
    
    chain = prompt | get_llm() | StrOutputParser()
    fixed_query = chain.invoke({
        "original_query": original_query,
        "error_message": error_message,
        "table_schema": table_schema
    })
    
    logging.info(f"Generated fixed query: {fixed_query}")
    return {
        "generated_query": fixed_query,
        "retry_count": state.get("retry_count", 0) + 1
    }